Decouple the Amount class from math performed on values

mwshortt opened the pull request:
* Remove IOU_SUFFIX from orderbookutils.js by using the correct currency and counterparty when available 
* Create new Values class that performs math without using the Amount class 
* Change serialize function in serializedtypes.js to accept values as well as amounts
* Clean up serializedtypes.js so it follows eslint rules

clark800 commented on src/core/value_XRP.js:
```diff
@@ -0,0 +1,30 @@
+
+Value_XRP.prototype.constructor = Value_XRP;
+
+Value_XRP.prototype._canonicalize = function(value) {
```
Nice polymorphism :+1:

clark800 commented on src/core/value_IOU.js:
```diff
@@ -0,0 +1,54 @@
+  DECIMAL_PLACES: 40
+});
+
+function Value_IOU(value) {
```
`IOUValue` and `XRPValue` sound a little more natural to me

wltsmrz replied:
Agree with this too

sublimator replied:
Yes

clark800 commented on src/core/value_IOU.js:
```diff
@@ -0,0 +1,54 @@
+Value_IOU.prototype.constructor = Value_IOU;
+
+Value_IOU.fromXRPValue = function(XRPvalue, bi_xns_unit) {
+  const newV = new Value_IOU(XRPvalue);
```
Can call it `newValue`?

clark800 commented on src/core/value_IOU.js:
```diff
@@ -0,0 +1,54 @@
+
+Value_IOU.prototype.multiply = function(multiplicand) {
+  let mult = multiplicand;
+  if (mult._XRP) {
```
I know this is more or less a result of the way rippled does math, but it's still really strange to have an `_XRP` flag in an IOU value class. Maybe instead of `fromXRPValue` we can just provide `multiplyByXRP` and `divideByXRP`?

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -71,7 +109,16 @@ OrderBookUtils.getOfferQuality = function(offer, currencyGets) {
   } else {
-    amount = Amount.from_json(offer.quality + IOU_SUFFIX);
+
+    let currency = offer.TakerGets.currency;
```
This can be done with `const`s only.

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -50,7 +85,10 @@ OrderBookUtils.getOfferTakerPaysFunded = function(offer) {
+  let currency = offer.TakerPays.currency;
+  let issuer = offer.TakerPays.issuer;
+
+  return Amount.from_json(createAmount(offer.TakerGets, currency, issuer));
```
Isn't `offer.TakerGets` an amount?

sublimator replied:
Could be either a string OR an object amount, assuming it's in rippled format, which the capitalisation of the word indicates. It seems this (seeming) bug existed prior to this patch. Note the diff:
`- return Amount.from_json(offer.TakerGets + IOU_SUFFIX);`

It seems like the util is used in `autobridgecalculator`, which is code I'm as yet unacquainted with.

sublimator replied:
Not sure if it's a bug, or just poorly named. Taking a peek: https://github.com/ripple/ripple-lib/blob/5b4deabd90e5237bd0165e65520f25456d32a410/src/core/autobridgecalculator.js#L123

sublimator replied:
Also see here:  https://github.com/ripple/ripple-lib/blob/5b4deabd90e5237bd0165e65520f25456d32a410/src/core/autobridgecalculator.js#L13

The code is used for `leg one` where TakerGets is supposed to be native, which makes sense I guess, where you want to trade some IOU for XRP, to then trade that XRP for the wanted IOU.

I guess that function needs some renaming / comments to make it clearer.

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -36,7 +62,16 @@ OrderBookUtils.getOfferTakerGetsFunded = function(offer) {
+  let currency = offer.TakerGets.currency;
+  let issuer = offer.TakerGets.issuer;
+
+  if (currency === undefined || issuer === undefined) {
```
Looks like this block of coded is somewhat redundant with the above block

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -1,19 +1,36 @@
+function createAmount(value, currency, counterparty) {
+  if (currency === undefined || counterparty === undefined) {
+    return {'value': value,
+      'currency': '000',
```
Still have to get rid of this hack

clark800 commented on src/core/amount.js:
```diff
@@ -148,20 +156,22 @@ Amount.prototype.subtract = function(subtrahend) {
+    new Value_XRP(multiplicandAmount._value, Amount.bi_xns_unit)
+    : new Value_IOU(multiplicandAmount._value);
+  const thisValue = new Value_IOU(this._value);
+  return this._copy(new BigNumber(thisValue.multiply(multiplyBy)._value));
```
The math methods in the Amount class should be simpler now that most of the logic is in the value classes. I don't think the amount class should have to create any new BigNumber or Value objects because the Value classes should return them. In fact, the Amount class should have no mention of BigNumber because the Value class handles all the math.

sublimator replied:
If you go down this path I agree with what Chris says above. If the amount class `has no mention of` of any BigNumber, it follows that the `._value` member should be one of these new *Value classes that encapsulate them.


Anyway, it just occurred to me, that there's no amount class for the new ripple api, that uses counterparty instead of issuer.

clark800 replied:
I don't think we need an amount class. The plan is to factor out `XRPValue` and `IOUValue` into a ripple-lib-extensions module. Client code can do math with these value objects. The only thing the Amount class has besides math is some parsing and formatting, which can be removed from ripple-lib and into a separate ripple-lib-extensions module (only needed by UI code). Of course it also keeps the currency and counterparty, but it doesn't actually track them properly for multiply and divide, so this isn't really useful. Then we could potentially replace the Amount class with just a plain javascript object: `{currency, counterparty, value}`.

sublimator replied:
> a plain javascript object: {currency, counterparty, value}

Composed entirely of primitives (strings) ? Or could I set the value as an IOUValue?

clark800 replied:
Most likely the `value` field would be an instance of IOUValue or XRPValue.

wltsmrz commented on src/core/orderbookutils.js:
```diff
@@ -1,19 +1,36 @@
+      'currency': '000',
+      'issuer': 'rrrrrrrrrrrrrrrrrrrrrhoLvTp'};
+  }
+  return {'value': value, 'currency': currency, 'issuer': counterparty};
```
"createAmount" sounds misleading. I'd suggest renaming it but it looks like we could probably just return a real Amount here rather than Amount.from_json()ing throughout the rest of the code

clark800 replied:
Good point, looks like everywhere createAmount is called it is wrapped in Amount.from_json, so we can just make this return an Amount object.

sublimator replied:
Or use IOUValue as you mentioned elsewhere.

wltsmrz commented on src/core/orderbook.js:
```diff
@@ -448,11 +449,10 @@ OrderBook.prototype.applyTransferRate = function(balance) {
-  .value;
+  const adjustedBalance = (new Value_IOU(balance))
+  .divide(new Value_IOU(this._issuerTransferRate))
+  .multiply(new Value_IOU(OrderBook.DEFAULT_TRANSFER_RATE))
```
Should probably just make _issuerTransferRate and OrderBook.DEFAULT_TRANSFER_RATE IOValues on assignment

wltsmrz commented on src/core/amount.js:
```diff
@@ -249,21 +259,21 @@ Amount.prototype.ratio_human = function(denominator, opts) {
  * for Ripple epoch.
  * @return {Amount} The product. Unit will be the same as the first factor.
  */
-Amount.prototype.product_human = function(factor, opts) {
```
`options = {}` in the method signature works, I think

wltsmrz commented on src/core/value.js:
```diff
@@ -0,0 +1,55 @@
+};
+
+Value.prototype.add = function(addend) {
+  let result = this._value.plus(addend._value);
```
Should probably `assert(addend instanceof Value)` or `if (!(addend instanceof Value)) myAddend = new Value(addend)`. I also think we should add accessors instead of ._value, but not important

clark800 replied:
Adding a flow annotation would be much better than those options.

wltsmrz replied:
How do you make Flow cast the argument to a Value?

clark800 replied:
You shouldn't automatically cast, that means the caller is erroneous in passing the wrong type. Flow can do the equivalent of the assert, but statically.

wltsmrz replied:
The caller is only erroneous if the function only accepts Values : )

clark800 replied:
Yeah, it keeps things a lot simpler if each function only accepts one type.

wltsmrz commented on src/core/value_IOU.js:
```diff
@@ -0,0 +1,54 @@
+  Value.call(this, value);
+}
+
+Value_IOU.prototype = Object.create(Value.prototype);
```
I guess it's ok but I would stick to convention

clark800 replied:
Speaking of which, how about we use ES6 classes for this?

wltsmrz replied:
Come to think of it this might be a good place to start using ES6 classes

wltsmrz replied:
Huh looks like we agree

sublimator commented on src/core/orderbookutils.js:
```diff
@@ -71,7 +109,16 @@ OrderBookUtils.getOfferQuality = function(offer, currencyGets) {
+    let issuer = offer.TakerGets.issuer;
+
+    if (currency === undefined || issuer === undefined) {
+      currency = offer.TakerPays.currency;
```
As always using IOU_SUFFIX before suggests, the issuer/currency don't matter at all. All this seems redundant when createAmount will set some neutral values anyway.

clark800 replied:
Well the initial purpose of this task was to eliminate the arbitrary values in IOU_SUFFIX.

`quality` is a ratio of amount values, so `getOfferQuality` should return an `IOUValue` rather than an `Amount`.

sublimator replied:
True, these are just as arbitrary and pointless, no matter that they are at least taken from one of TakerPays / TakerGets.

mwshortt commented:
Thank you for all of your comments! Here is the updated version

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,43 @@
+    super(value);
+  }
+
+  multiplyByXRP(multiplicand: {_value: BigNumber, _bi_xns_unit: number}) {
```
I think we should be able to make the type of multiplicand be `XRPValue`.

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,43 @@
+  multiplyByXRP(multiplicand: {_value: BigNumber, _bi_xns_unit: number}) {
+    const constant = new BigNumber((multiplicand._bi_xns_unit).toString());
+    const value = new BigNumber(multiplicand._value);
+    multiplicand._value = (value).times(constant);
```
Instead of mutating I think we could create a new `IOUValue` instance.

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,43 @@
+
+  equals(comparator) {
+    return (comparator instanceof IOUValue)
+    && this._value.equals(comparator.value);
```
One is `_value` and one is `value`, is that right?

clark800 commented on src/core/XRPValue.js:
```diff
@@ -0,0 +1,29 @@
+
+  constructor(value, bi_xns_unit) {
+    super(value);
+    this._bi_xns_unit = bi_xns_unit;
```
Might be worth considering a rename of this property.

clark800 commented on src/core/amount.js:
```diff
@@ -3,22 +3,19 @@
 
 function inverse(number) {
-  return (new BigNumber(number)).toPower(-1);
+  return new IOUValue(number).invert()._value;
```
Property names that start with underscore are supposed to be private.

clark800 commented on src/core/amount.js:
```diff
@@ -38,25 +35,25 @@ function Amount() {
-  bi_xns_max: new BigNumber(1e17),
-  bi_xns_min: new BigNumber(-1e17),
-  bi_xns_unit: new BigNumber(1e6),
+  bi_5: new Value(5)._value,
```
I don't know if anyone is using these, but now might be a good time to delete them. We shouldn't be exposing the bignumber implementation here.

mwshortt replied:
Most of them aren't being used, but a handful of them still are (bi_man_max_value, bi_man_min_value, bi_xns_max, bi_xns_min, bi_xns_unit). I've removed the superfluous ones

clark800 replied:
For the constants that we need, I think they should be local variables in the appropriate module. I think the ones you listed probably belong in the `XRPValue` or `IOUValue` classes.

clark800 commented on src/core/amount.js:
```diff
@@ -113,32 +110,45 @@ Amount.is_valid_full = function(j) {
 
-  return this._copy(this._value.plus(addendAmount._value));
+  const thisValue = new IOUValue(this._value);
+  const addendValue = new IOUValue(addendAmount._value);
```
You can just make `this._value` be stored as an instance of `XRPValue` or `IOUValue` so you don't have to do this every time.

clark800 commented on src/core/amount.js:
```diff
@@ -249,21 +264,20 @@ Amount.prototype.ratio_human = function(denominator, opts) {
  */
-Amount.prototype.product_human = function(factor, opts) {
-  opts = opts || {};
+Amount.prototype.product_human = function(fac, options = {}) {
```
Better not to use abbreviations, especially in method signatures.

clark800 commented on src/core/amount.js:
```diff
@@ -343,12 +358,15 @@ Amount.prototype.invert = function() {
  */
 
 Amount.prototype.canonicalize = function(roundingMode) {
```
Does Amount even need a canonicalize method anymore? Don't the value classes already ensure canonicalization?

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -1,19 +1,61 @@
+  if (!currency) {
+    currency = offer.TakerGets.currency;
+  }
+  return currency;
```
`return offer.TakerPays.currency || offer.TakerGets.currency`

clark800 commented on src/core/amount.js:
```diff
@@ -113,32 +110,45 @@ Amount.is_valid_full = function(j) {
 };
 
 // be sure that _is_native is set properly BEFORE calling _set_value
 Amount.prototype._set_value = function(value, roundingMode) {
```
We should be able to get rid of the `_set_value` method by just requiring that a Value instance is passed into the Amount constructor.

sublimator commented on bin/ci.sh:
```diff
@@ -11,7 +11,7 @@ typecheck() {
   REPO_URL="https://raw.githubusercontent.com/ripple/javascript-style-guide"
   curl "$REPO_URL/es6/eslintrc" > ./eslintrc
-  echo "parser: esprima-fb" >> ./eslintrc
+  echo "parser: babel-eslint" >> ./eslintrc
```
This supports facebook flow (vaguely recall someone telling me that's what esprima-fb was for )  
?

sublimator replied:
I guess so, judging by the rest of the code

sublimator commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+
+  getMantissaDecimalString() {
+    let mantissa = this._value.abs().toPrecision(16)
+    .replace(/\./, '')      // remove decimal point
```
Is this non indenting subsequent lines of multi-line expressions in our style-guide or something? I've seen it a bit recently.

clark800 replied:
I prefer indentation.

mwshortt commented:
Updated once again:
* In the Amount class, this._value now stores an IOU or XRP Value
* In IOU and XRP Value, this._value is now private
* Out BN implementation is now hidden (I removed the exposed BN constants in the Amount class)

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+  multiply(multiplicand: Value) {
+    const val = this._value;
+    const mult = multiplicand._value;
+    const result = (val).times(mult);
```
Seems more consistent to inline like the others: `const result = this._value.times(multiplicand._value);`

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+    return this._canonicalize(result);
+  }
+
+  scale(scaleFactor: Value) {
```
This function can be deleted since it's exactly the same as multiply now.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+    return this._canonicalize(result);
+  }
+
+  add(addend: Value) {
```
Throw an exception if types aren't the same in `add` and `subtract`

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+
+  divide(divisor: Value) {
+    if (this._value.isNaN()) {
+      throw new Error('Invalid dividend');
```
I'm not sure why we are checking for NaN only in division. Probably we should either delete these checks or maybe check for NaN in canonicalize?

clark800 commented on src/core/amount.js:
```diff
@@ -729,7 +703,8 @@ Amount.prototype.parse_native = function(j) {
 Amount.prototype.parse_value = function(j) {
   this._is_native = false;
-  this._set_value(new BigNumber(j), BigNumber.ROUND_DOWN);
+  const newValue = new IOUValue(j).toPrecision(16, Value.getBNRoundDown());
```
`toPrecision` sounds like it should return a string since that's what bignumber.js toPrecision returns. Maybe a better option would be to pass the rounding mode to the IOUValue constructor and call canonicalize from the constructor.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+    return this;
+  }
+
+  toPrecision(significantDigits, roundingMode) {
```
Can remove this method.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+    return this._value.e;
+  }
+
+  getMantissaDecimalString() {
```
Can this function be moved to serializedtypes.js by using `value.abs().toString()` ?

mwshortt replied:
I changed this to now call utils.js' method of the same name, passing this._value.abs()

clark800 replied:
I'd like if we can delete this method entirely because it seems too low-level to be on the value class.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  equals(comparator) {
+    return (comparator instanceof Value)
```
I think an XRPValue and IOUValue should be not equal regardless of their numerical value.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  greaterThan(comparator: Value) {
+    return this._value.greaterThan(comparator._value);
```
Comparisons should probably throw an exception if you try to compare different subtypes.

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,44 @@
+    return super.divide(divisor);
+  }
+
+  negated() {
```
For consistency, we should rename to `negate`

clark800 commented on src/core/amount.js:
```diff
@@ -113,32 +100,37 @@ Amount.is_valid_full = function(j) {
-  this._value = value.isZero() && value.isNegative() ? value.negated() : value;
-  this.canonicalize(roundingMode);
+Amount.prototype._set_value = function(value) {
+  assert(value instanceof Value);
```
Might be good to use a flow annotation instead of the assert.

clark800 commented on src/core/orderbookutils.js:
```diff
@@ -1,19 +1,56 @@
+  if (!issuer) {
+    issuer = offer.TakerGets.issuer;
+  }
+  return issuer;
```
Like above: `offer.TakerPays.issuer || offer.TakerGets.issuer;`

clark800 commented on src/core/XRPValue.js:
```diff
@@ -0,0 +1,43 @@
+
+  constructor(value) {
+    super(value);
+    this.ripple_units = new BigNumber(1e6);
```
I think we're trying to use camelCase now.

clark800 commented on src/core/XRPValue.js:
```diff
@@ -0,0 +1,43 @@
+
+class XRPValue extends Value {
+
+  constructor(value) {
```
It might be a good idea to throw an exception in the constructor if `value` has more than 6 digits of precision past the decimal point. This would help catch bugs where an IOUValue is being cast to an XRPValue.

mwshortt replied:
That's a good idea, but there are four tests in amount-tests.js using Amount.from_quality that violate this rule even though they are using XRP.

clark800 replied:
Since quality should always be an IOUValue, I think that won't be an issue.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+  DECIMAL_PLACES: 40
+});
+
+class Value {
```
We should make this an abstract base class so that there are only two Value classes that appear in the code: in the constructor, throw an exception if `new.target === Value`. Also, in all the places that instantiate Value directly, replace it with either XRPValue or IOUValue.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,134 @@
+    this._value = new BigNumber(value, base);
+  }
+
+  static getBNRoundDown() {
```
Might be simpler to use class properties like the BigNumber class does. Basically just `Value.ROUND_DOWN = BigNumber.ROUND_DOWN` etc.

clark800 commented on src/core/amount.js:
```diff
@@ -614,10 +586,12 @@ function(quality, counterCurrency, counterIssuer, opts) {
-  var nativeAdjusted = adjusted;
+  const valueStr = mantissa.toString() + 'e' + offset.toString();
+  let nativeAdjusted = this._is_native ?
+    new XRPValue(valueStr) : new IOUValue(valueStr);
```
"quality" is a ratio, so it should always be an `IOUValue`

mwshortt replied:
But the tests are expecting canonicalization according to XRPValue's canonicalize function when the currency is native. This results in assertion errors where it is expecting the rounded value: 
```AssertionError: '90,442.19667786521/XRP' === '90,442.196677/XRP'```

clark800 replied:
Hmmm, this is another quirk of the prior design: it doesn't really make sense to convert a quality to an amount because a quality is a ratio and an amount doesn't support ratio dimensions.

There are a couple options: throw an exception when from_quality's second parameter is XRP (this would be a breaking change). Or we could put in a hack that uses IOUValue.toFixed to cast to an XRPValue without violating the precision constraint. We should try to avoid breaking changes since this is a refactoring pull request though. Maybe later we can do the breaking change in a separate pull request. Even better would be to just delete Amount.from_quality and replace it with a function (not in amount.js) called `parseQuality` that returns an IOUValue.

mwshortt replied:
Even if we use this hack we still run into issues with _check_limits (called from _set_value), because it's then trying to compare XRPValues with the IOUValue created in parse_quality. 
What about initially having it be an IOUValue and only at the end, when _set_value is called, turning it into an XRPValue if it's native? This will allow the >6 decimal place check in the XRPValue constructor to work appropriately

clark800 replied:
Yeah, that's what I was thinking.

mwshortt replied:
@clark800 Re your most recent comment: this is why I am passing an IOUValue instance to the constructor of XRPValue.

clark800 replied:
The signature of the XRPValue constructor is `constructor(value: string | BigNumber)`, so this should be a type error, but I guess flow isn't smart enough to catch it? I think if you just add a `.toString()` it should work.

mwshortt replied:
It does, will fix

mwshortt commented:
Integrated @clark800 's comments to make Value.js an abstract base class and other fixes. Let me know how it looks!

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,54 @@
+
+  constructor(value: string | BigNumber, base: ?number) {
+    super(value, base);
+    this._type = IOUValue;
```
Can use `instance.constructor`

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,54 @@
+    if (value.isNaN()) {
+      throw new Error('Invalid result');
+    }
+    return new IOUValue(value.toPrecision(16), undefined);
```
Can we remove all the `undefined`s?

mwshortt replied:
I was getting flow errors if I tried to call a function without all the variables, so I added them in. But I can add a default for base in the IOUValue constructor and that solves the flow errors.

clark800 replied:
That sounds like a good option

clark800 commented on src/core/IOUValue.js:
```diff
@@ -0,0 +1,54 @@
+
+  equals(comparator) {
+    return (comparator instanceof IOUValue)
+    && this._value.equals(comparator._value);
```
Indentation

clark800 commented on src/core/XRPValue.js:
```diff
@@ -0,0 +1,60 @@
+
+  equals(comparator) {
+    return (comparator instanceof XRPValue)
+    && this._value.equals(comparator._value);
```
Indentation

clark800 commented on src/core/utils.js:
```diff
@@ -1,8 +1,9 @@
-function getMantissaDecimalString(bignum) {
-  let mantissa = bignum.toPrecision(16)
-    .replace(/\./, '')      // remove decimal point
+// Should only be used directly when the string is from a BigNumber
```
I don't think it matters if the string is from a BigNumber. Parameter name could be `decimalString` maybe? Any decimal-encoded numeric string should be the same right? Also, I don't think the function requires 16 digits of precision. It might be good to add a case for if `mantissa.length` is greater than 16 though so that it works for all inputs.

mwshortt replied:
You're correct. If the ```mantissa.length``` is greater than 16 should it return it or should it add trailing zeros to a different length (maybe adding a parameter to the function?)?

clark800 replied:
I think for our use cases we would probably want to truncate the least significant digits. We might want to add "16" to the name of the function as a reminder.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,121 @@
+class Value {
+
+  constructor(value: string | BigNumber, base: number) {
+    if (this.constructor.name === 'Value') {
```
Does `this.constructor === Value` work?

mwshortt replied:
It does, I will simplify that

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  subtract(subtrahend: Value) {
+    if (this._type !== subtrahend._type) {
```
Could save two lines per method if we convert these to asserts.

clark800 commented on src/core/amount.js:
```diff
@@ -729,7 +705,9 @@ Amount.prototype.parse_native = function(j) {
   this._is_native = false;
-  this._set_value(new BigNumber(j), BigNumber.ROUND_DOWN);
+  const newValue = new IOUValue(
+    new IOUValue(j).toPrecision(16, Value.getBNRoundDown()));
```
Looks like the suggestion for this got lost.

mwshortt replied:
It seems messier to me to be passing another value into the constructor for one specific case versus calling a specific function. But I can change it

clark800 replied:
Yeah, the messiness is the fault of the legacy code that uses two different rounding modes.

The advantage is that we don't have to expose the toPrecision method on the interface. So it's a trade-off between an option on the constructor, vs. an additional method on the interface. I was thinking an option on one method counts as less complexity than a new method, but it's debatable. The other factor is that toPrecision seems less of something that a "Value" should do than having options for how to construct one. What do you think?

mwshortt replied:
Yeah it's a close call and I see your point. The BigNumber class does have the toPrecision function so it seems plausible that the Value could do it too. I'm going to go with the constructor option as I agree it's better to not expose the toPrecision method.

clark800 commented on src/core/value.js:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  round(decimalPlaces: number, roundingMode: number) {
+    this._value = this._value.round(decimalPlaces, roundingMode);
```
Can we just return the new value like the other methods and avoid mutation?

mwshortt commented:
Updated once again

clark800 commented on src/core/utils.js:
```diff
@@ -1,8 +1,9 @@
+  let mantissa = decimalString.replace(/\./, '')      // remove decimal point
     .replace(/e.*/, '')     // remove scientific notation
     .replace(/^0*/, '');    // remove leading zeroes
   while (mantissa.length < 16) {
```
What about when length > 16?

mwshortt replied:
Oh I get what you meant earlier, use a toFixed() method to round it to only be of length 16, right?

clark800 commented on src/core/XRPValue.js:
```diff
@@ -0,0 +1,59 @@
+
+  constructor(value: string | BigNumber) {
+    super(value);
+    this.rippleUnits = new BigNumber(1e6);
```
This can be a global constant in this module and IOUValue.js, which would actually reduce it's exposure. Also I just noticed, we should probably stick to the file naming convention of all lowercase, no separator between words.

mwshortt replied:
Okay. So rename to xrpvalue and iouvalue?

clark800 replied:
yeah

mwshortt commented:
Updated to include the newest round of comments

clark800 commented on src/core/amount.js:
```diff
@@ -626,22 +598,26 @@ function(quality, counterCurrency, counterIssuer, opts) {
   }
+  if (this._is_native) {
+    this._set_value(
+      new XRPValue(nativeAdjusted.round(6, Value.getBNRoundDown())));
```
It looks like this is passing an IOUValue instance to the constructor of XRPValue.

clark800 commented on src/core/utils.js:
```diff
@@ -1,16 +1,25 @@
     .replace(/e.*/, '')     // remove scientific notation
     .replace(/^0*/, '');    // remove leading zeroes
+  if (mantissa.length > 16) {
+    const rounded = Number(mantissa).toFixed(16);
```
This isn't safe because the maximum javascript integer is 2^53-1 = 9007199254740991 < 10^16-1.

mwshortt replied:
Right. Instead should I create a new IOUValue and or throw an error if the mantissa is longer than 16 or is there a better solution I'm missing?

clark800 replied:
You can truncate by doing `mantissa.substring(0, 16)`

mwshortt replied:
Hmm okay - so it's alright to truncate versus treating it like a number and rounding it somehow?

clark800 replied:
Well truncate is equivalent to round down. You could argue that we should use the same rounding mode as the rest of the program, which is round half up. But because this function is basically getting the first 16 digits of the mantissa, round down seems more natural to me. It could go either way though.

mwshortt commented:
Updated with the two small fixes

clark800 commented on src/core/amount.js:
```diff
@@ -38,25 +31,17 @@ function Amount() {
-  bi_xns_max: new BigNumber(1e17),
-  bi_xns_min: new BigNumber(-1e17),
-  bi_xns_unit: new BigNumber(1e6),
+  // man refers to mantissa
```
This is good

clark800 commented:
LGTM!!
